changes.go

```diff
@@ -65,6 +66,31 @@ func (cs *changeset) add(method string, requires []string, args ...interface{})
 	return change
 }
 
+// sorted returns the changes sorted by requirements, required first.
+func (cs *changeset) sorted() []*Change {
+	numChanges := len(cs.changes)
+	records := make(map[string]bool, numChanges)
+	sorted := make([]*Change, 0, numChanges)
+	changes := make([]*Change, numChanges, numChanges*2)
+	copy(changes, cs.changes)
+mainloop:
+	for len(changes) != 0 {
+		// Note that all valid bundles have at least two changes
+		// (add one charm and deploy one service).
+		change := changes[0]
+		changes = changes[1:]
+		for _, r := range change.Requires {
```

Comment:
We already start with a slice, so I've found quite straightforward to use reslicing rather than channels here.